### PR TITLE
BAH-1856 | Saving diagnostic report against an open order

### PR DESCRIPTION
**Pull request by Arjun-Go**
## Requirements

- [x] This PR has a proper title that briefly describes the work done
- [x] I have squashed / amended the comments to make it more readable
- [x] I have included link to all the JIRA ticket('s)
- [x] I wrote the code as a pair or atleast performed a self-review of my own code
- [x] I have updated the documentation on [Bahmni Wiki](https://bahmni.atlassian.net/wiki/spaces/BAH/overview) or [README](https://github.com/Bahmni/bahmni-lab-frontend/blob/main/README.md) (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Summary
Enhance POST call to save diagnostic report against open order

## Screenshots
image

## JIRA tickets
https://bahmni.atlassian.net/jira/software/c/projects/BAH/boards/35?modal=detail&selectedIssue=BAH-1856

### CODE REVIEW

File: src/patient-lab-details/patient-lab-details.test.tsx [hunk 1]
```diff
@@ -334,6 +334,111 @@ describe('Patient lab details', () => {
       ).toBeInTheDocument(),
     )
   })
+
+  it('should populate based on property if a pending order is selected', async () => {
```
Comment (N0-man) [inline src/patient-lab-details/patient-lab-details.test.tsx:287]: this test is too long to make sense

File: src/patient-lab-details/patient-lab-details.test.tsx [hunk 2]
```diff
@@ -334,6 +334,111 @@ describe('Patient lab details', () => {
...
+    Object.defineProperty(window, 'localStorage', {value: localStorageMock})
+    when(usePatient)
```
Comment (N0-man) [inline src/patient-lab-details/patient-lab-details.test.tsx]: wrap this into a function as `usePatientMock()`
similarly you can use `mockLocalStorage()` reduce the noisy lines

File: src/patient-lab-details/patient-lab-details.test.tsx [hunk 3]
```diff
@@ -334,6 +334,111 @@ describe('Patient lab details', () => {
...
+
+    const file = new File(['content'], 'test.pdf', {type: 'application/pdf'})
+    const mockedOpenmrsFetch = openmrsFetch as jest.Mock
+    mockedOpenmrsFetch
+      .mockReturnValueOnce(mockPendingLabOrdersResponse)
+      .mockReturnValueOnce(mockLabTestsResponse)
+      .mockReturnValueOnce(mockUploadFileResponse)
+      .mockReturnValue(mockDiagnosticReportResponse)
+
+    const mockedLayout = useLayoutType as jest.Mock
```
Comment (N0-man) [inline src/patient-lab-details/patient-lab-details.test.tsx]: use `@openmrs/esm-framework/mock` so you dont need to mock most of the esm-framework functions

File: src/patient-lab-details/patient-lab-details.test.tsx [hunk 4]
```diff
@@ -334,6 +334,111 @@ describe('Patient lab details', () => {
...
+
+    const fileInput = screen.getByLabelText(
+      'Drag and drop files here or click to upload',
+    ) as HTMLInputElement
+
+    uploadFiles(fileInput, [file])
+
+    expect(fileInput.files.length).toBe(1)
+    const fileName = await screen.findByText('test.pdf')
+    expect(fileName).toBeInTheDocument()
```
Comment (N0-man) [inline src/patient-lab-details/patient-lab-details.test.tsx]: can you break this down into functions to make sense on the user behavior 

e.g. 

```
async function verifyFileName(fileInput: HTMLInputElement) {
    expect(fileInput.files.length).toBe(1)
    const fileName = await screen.findByText('test.pdf')
    expect(fileName).toBeInTheDocument()
  }
```
and 

```
async function saveReport() {
    const saveButton = screen.getByRole('button', {name: /save and upload/i})

    expect(saveButton).not.toBeDisabled()
    userEvent.click(saveButton)

    await waitFor(() =>
      expect(
        screen.getByText(/Report successfully uploaded/i),
      ).toBeInTheDocument(),
    )
  }
```

and then your test would look something like this...
```
...
    uploadFiles(fileInput)
    await verifyFileName(fileInput)
    await saveReport()
```

File: src/report-table/report-table.component.tsx [hunk 5]
```diff
@@ -27,9 +27,9 @@ const ReportTable = ({patientUuid}) => {
     Error
   >(getReportTableDataURL(patientUuid), fetcher)
 
-  if (reports?.data) {
-    reports.data = sampleResponse.data
-  }
+  // if (reports?.data) {
+  //   reports.data = sampleResponse.data
+  // }
```
Comment (N0-man) [inline src/report-table/report-table.component.tsx]: remove it if you dont need it

File: src/upload-report/upload-report.tsx [hunk 6]
```diff
@@ -58,6 +64,18 @@ const UploadReport: React.FC<UploadReportProps> = ({
       )
       if (uploadFileResponse.ok) {
         const url = uploadFileResponse.data.url
+        const isPendingOrderInPayload =
```
Comment (N0-man) [inline src/upload-report/upload-report.tsx:67]: SRP violation

Comment (N0-man) [inline src/upload-report/upload-report.tsx]: SRP violation

File: src/patient-lab-details/patient-lab-details.test.tsx (hunk 2 above)
Comment (Arjun-Go) [reply src/patient-lab-details/patient-lab-details.test.tsx]: We have moved these functions to the `beforeEach`

File: src/patient-lab-details/patient-lab-details.test.tsx (hunk 3 above)
Comment (Arjun-Go) [reply src/patient-lab-details/patient-lab-details.test.tsx]: updated the test file to use `@openmrs/esm-framework/mock`

File: src/patient-lab-details/patient-lab-details.test.tsx (hunk 4 above)
Comment (Arjun-Go) [reply src/patient-lab-details/patient-lab-details.test.tsx]: refactored test according to the user actions

File: src/report-table/report-table.component.tsx (hunk 5 above)
Comment (Arjun-Go) [reply src/report-table/report-table.component.tsx]: sure

File: src/patient-lab-details/patient-lab-details.test.tsx (hunk 1 above)
Comment (Arjun-Go) [reply src/patient-lab-details/patient-lab-details.test.tsx:287]: the test has been refactored

File: src/upload-report/upload-report.tsx (hunk 6 above)
Comment (Arjun-Go) [reply src/upload-report/upload-report.tsx]: has been taken care of in the latest commits

**Pull request merged**